File: apache-s4-0.6.0/subprojects/s4-core/src/main/java/org/apache/s4/core/monitor/MarkovChain.java
package org.apache.s4.core.monitor;

import java.text.DecimalFormat;

public class MarkovChain {
	private double transitionMatrix[][];
	private double prediction[];

	/**
	 * Inicialización de la matriz de transición y el arreglo para la
	 * distribución estacionaria
	 */
	public MarkovChain() {
		setTransitionMatrix(new double[3][3]);
		setPrediction(new double[3]);
	}

	/**
	 * Cálculo para la distribución estacionaria, para esto se analizará la tasa
	 * de procesamiento respecto a un período t y t+1, de esta manera se
	 * verifica a que estado a variado. Por lo que se consideraron tres posibles
	 * estados en la Cadena de Markov: Ocioso, Estable e Inestable. Ocioso se
	 * refiere a cuando existen una cantidad de servicios que poseen un alto
	 * tiempo de ocio. Estable es cuando se posee con un buen número de
	 * procesamiento. E inestable cuando existe un tasa de procesamiento menor
	 * que la tasa de llegada, por lo que generará colas en el sistema.
	 * 
	 * @param rho Tasa de procesamiento del sistema en un 
	 * @param n
	 * @param iteration
	 * @return
	 */

	public double[] calculatePrediction(double[] rho, int n, int iteration) {
		int cont[] = new int[3];

		for (int i = 0; i < n - 1; i++) {
			if ((rho[i] < 0.5) && (rho[i + 1] < 0.5)) {
				transitionMatrix[0][0]++;
				cont[0]++;
			} else if ((rho[i] < 0.5) && (rho[i + 1] >= 0.5)
					&& (rho[i + 1] <= 1.5)) {
				transitionMatrix[0][1]++;
				cont[0]++;
			} else if ((rho[i] < 0.5) && (rho[i + 1] > 1.5)) {
				transitionMatrix[0][2]++;
				cont[0]++;
			} else if ((rho[i] >= 0.5) && (rho[i] <= 1.5) && (rho[i + 1] < 0.5)) {
				transitionMatrix[1][0]++;
				cont[1]++;
			} else if ((rho[i] >= 0.5) && (rho[i] <= 1.5)
					&& (rho[i + 1] >= 0.5) && (rho[i + 1] <= 1.5)) {
				transitionMatrix[1][1]++;
				cont[1]++;
			} else if ((rho[i] >= 0.5) && (rho[i] <= 1.5) && (rho[i + 1] > 1.5)) {
				transitionMatrix[1][2]++;
				cont[1]++;
			} else if ((rho[i] > 1.5) && (rho[i + 1] < 0.5)) {
				transitionMatrix[2][0]++;
				cont[2]++;
			} else if ((rho[i] > 1.5) && (rho[i + 1] >= 0.5)
					&& (rho[i + 1] <= 1.5)) {
				transitionMatrix[2][1]++;
				cont[2]++;
			} else if ((rho[i] > 1.5) && (rho[i + 1] > 1.5)) {
				transitionMatrix[2][2]++;
				cont[2]++;
			}
		}

		for (int i = 0; i < 3; i++) {
			if (cont[i] != 0) {
				for (int j = 0; j < 3; j++) {
					transitionMatrix[i][j] /= cont[i];
				}
			}
		}

		int acum;
		int i = 0;

		for (int k = 0; k < 3; k++) {

			acum = 0;
			for (int j = 0; j < 3; j++) {
				if (transitionMatrix[k][j] == 0) {
					acum++;
				}
			}

			if (acum != 3) {
				i = k;
				break;
			}

		}

		double u;
		double probAcum;
		cont = new int[3];

		for (int k = 0; k < iteration; k++) {

			u = Math.random();
			probAcum = 0;
			for (int j = 0; j < 3; j++) {
				probAcum += transitionMatrix[i][j];
				if (u <= probAcum) {
					cont[j]++;
					i = j;
					break;
				}
			}

		}

		for (int k = 0; k < 3; k++) {
			prediction[k] = (double) cont[k] / (double) iteration;
		}

		return prediction;
	}

	public double[][] getTransitionMatrix() {
		return transitionMatrix;
	}

	public void setTransitionMatrix(double transitionMatrix[][]) {
		this.transitionMatrix = transitionMatrix;
	}

	public double[] getPrediction() {
		return prediction;
	}

	public void setPrediction(double prediction[]) {
		this.prediction = prediction;
	}

	public String showPrediction() {
		DecimalFormat decimalFormat = new DecimalFormat("0.000");
		return "[ " + decimalFormat.format(prediction[0]) + " "
				+ decimalFormat.format(prediction[1]) + " "
				+ decimalFormat.format(prediction[2]) + " ]";
	}

	@Override
	public String toString() {
		DecimalFormat decimalFormat = new DecimalFormat("0.000");

		String transitionMatrixString = new String("╔═══════════════════╗\n║ "
				+ decimalFormat.format(this.transitionMatrix[0][0]) + " "
				+ decimalFormat.format(this.transitionMatrix[0][1]) + " "
				+ decimalFormat.format(this.transitionMatrix[0][2]) + " ║\n║ "
				+ decimalFormat.format(this.transitionMatrix[1][0]) + " "
				+ decimalFormat.format(this.transitionMatrix[1][1]) + " "
				+ decimalFormat.format(this.transitionMatrix[1][2]) + " ║\n║ "
				+ decimalFormat.format(this.transitionMatrix[2][0]) + " "
				+ decimalFormat.format(this.transitionMatrix[2][1]) + " "
				+ decimalFormat.format(this.transitionMatrix[2][2])
				+ " ║\n╚═══════════════════╝");
		return transitionMatrixString;
	}
}

File: misc/MarkovChain/MarkovChain.m
function [Prediction] = MarkovChain(Lambda,Mu,N,Iteration)

    % λ = lambda
    % num = [1 N];
    % Tl = normrnd(1000,500,num);
    % Tl = poissrnd(2000,num);
    Tl = Lambda;

    for i = 1:N
        if( Tl(i) < 0 )
            Tl(i) = 0;
        end
    end

    % μ = mu
    % num = [1 N];
    % Ts = normrnd(1000,250,num);
    Ts = Mu;

    for i = 1:N
        if( Ts(i) < 0 )
            Ts(i) = 0;
        end
    end

    % Cola del sistema
    C = [];

    C(1) = Tl(1) - Ts(1);
    if(C(1) < 0)
        C(1) = 0;
    end

    for i=2:N
        C(i) = C(i-1) + (Tl(i) - Ts(i));
        if(C(i) < 0)
            C(i) = 0;
        end
    end


    % Tasa de procesamiento
    Tp = Tl ./ Ts;
    for i=1:N
        if(Tp(i) == Inf)
            Tp(i) = 0;
        end
    end

    % Cadena de Markov
    P = zeros(3);
    Cont = [0 0 0];

    for i=1:(N-1)
        %Analisis de no-replicacion
        if (Tp(i) < 0.5) && (Tp(i+1) < 0.5)
            P(1,1) = P(1,1) + 1;
            Cont(1) = Cont(1) + 1;
        elseif (Tp(i) < 0.5) && (Tp(i+1) >= 0.5) && (Tp(i+1) <= 1.5)
            P(1,2) = P(1,2) + 1;
            Cont(1) = Cont(1) + 1;
        elseif (Tp(i) < 0.5) && (Tp(i+1) > 1.5)
            P(1,3) = P(1,3) + 1;
            Cont(1) = Cont(1) + 1;
        %Analisis de estable    
        elseif (Tp(i) >= 0.5) && (Tp(i) <= 1.5) && (Tp(i+1) < 0.5)
            P(2,1) = P(2,1) + 1;
            Cont(2) = Cont(2) + 1;
        elseif (Tp(i) >= 0.5) && (Tp(i) <= 1.5) && (Tp(i+1) >= 0.5) && (Tp(i+1) <= 1.5)
            P(2,2) = P(2,2) + 1;
            Cont(2) = Cont(2) + 1;
        elseif (Tp(i) >= 0.5) && (Tp(i) <= 1.5) && (Tp(i+1) > 1.5)
            P(2,3) = P(2,3) + 1;
            Cont(2) = Cont(2) + 1;
        %Analisis de replicacion    
        elseif (Tp(i) > 1.5) && (Tp(i) < 0.5)
            P(3,1) = P(3,1) + 1;
            Cont(3) = Cont(3) + 1;
        elseif (Tp(i) > 1.5) && (Tp(i+1) >= 0.5) && (Tp(i+1) <= 1.5)
            P(3,2) = P(3,2) + 1;
            Cont(3) = Cont(3) + 1;
        else
            P(3,3) = P(3,3) + 1;
            Cont(3) = Cont(3) + 1;
        end

    end

    % Normalizar las probabilidades
    if(Cont(1) ~= 0)
        P(1,:) = P(1,:)./Cont(1);
    end

    if(Cont(2) ~= 0)
        P(2,:) = P(2,:)./Cont(2);
    end

    if(Cont(3) ~= 0)
        P(3,:) = P(3,:)./Cont(3);
    end

    C = [0 0 0];

    for k=1:3
        if(~isequal(P(k,:),C))
            i = k;
            break;
        end
    end

    for n = 1:Iteration
        u = rand;
        probAcum = 0;
        for j=1:3
            probAcum = probAcum + P(i,j);
            if u <= probAcum
                C(j)=C(j)+1;
                i = j;
                break
            end
        end
    end

    Prediction = C/Iteration;

end